packages/dd-trace/src/opentelemetry/metrics/periodic_metric_reader.js

```diff
@@ -0,0 +1,353 @@
+    if (!this.#observableInstruments.includes(instrument)) {
+      this.#observableInstruments.push(instrument)
+    }
+  }
+
+  /**
+   * Forces an immediate collection and export of all metrics.
+   * @returns {void}
+   */
+  forceFlush () {
+    this.#collectAndExport()
+  }
+
+  /**
+   * Shuts down the reader and stops periodic collection.
+   * @returns {void}
+   */
+  shutdown () {
+    this.#clearTimer()
+    this.forceFlush()
+  }
+
+  #startTimer () {
+    if (this.#timer) return
+
+    this.#timer = setInterval(() => {
+      this.#collectAndExport()
+    }, this.#exportInterval).unref()
+  }
+
+  #clearTimer () {
+    if (this.#timer) {
+      clearInterval(this.#timer)
+      this.#timer = null
+    }
+  }
+
+  #collectAndExport (callback = () => {}) {
+    const allMeasurements = this.#measurements.splice(0)
+
+    for (const instrument of this.#observableInstruments) {
+      allMeasurements.push(...instrument.collect())
+    }
+
+    if (allMeasurements.length === 0) {
+      callback()
+      return
+    }
+
+    const metrics = this.#aggregator.aggregate(
+      allMeasurements,
+      this.#cumulativeState,
+      this.#lastExportedState
+    )
+
+    this.exporter.export(metrics, callback)
+  }
+}
+
+/**
+ * MetricAggregator aggregates individual measurements into metric data points.
+ * @private
+ */
+class MetricAggregator {
+  #temporalityPreference
+  #startTime
+
+  constructor (temporalityPreference = TEMPORALITY.DELTA) {
+    this.#temporalityPreference = temporalityPreference
+    this.#startTime = Number(process.hrtime.bigint())
+  }
+
+  #getTemporality (type) {
+    // UpDownCounter and Observable UpDownCounter always use CUMULATIVE
+    if (type === METRIC_TYPES.UPDOWNCOUNTER || type === METRIC_TYPES.OBSERVABLEUPDOWNCOUNTER) {
+      return TEMPORALITY.CUMULATIVE
+    }
+
+    // Gauge always uses last-value aggregation
+    if (type === METRIC_TYPES.GAUGE) {
+      return TEMPORALITY.GAUGE
+    }
+
+    switch (this.#temporalityPreference) {
+      case TEMPORALITY.CUMULATIVE:
+        return TEMPORALITY.CUMULATIVE
+      case TEMPORALITY.LOWMEMORY:
+        // LOWMEMORY: only synchronous Counter and Histogram use DELTA, Observable Counter uses CUMULATIVE
+        return (type === METRIC_TYPES.COUNTER || type === METRIC_TYPES.HISTOGRAM)
+          ? TEMPORALITY.DELTA
+          : TEMPORALITY.CUMULATIVE
+      default:
+        return TEMPORALITY.DELTA
+    }
+  }
+
+  aggregate (measurements, cumulativeState, lastExportedState) {
+    const metricsMap = new Map()
+
+    for (const measurement of measurements) {
+      const {
+        name,
+        description,
+        unit,
+        type,
+        instrumentationScope,
+        value,
+        attributes,
+        timestamp
+      } = measurement
+
+      const scopeKey = this.#getScopeKey(instrumentationScope)
+      const metricKey = `${scopeKey}:${name}:${type}`
+      const attrKey = JSON.stringify(attributes)
```

Comment:
Using the result of JSON.stringify on an object as key is not stable.

E.g., { a: 1, b: 2 } is not equal to { b: 2, a: 1 }

You can use a library such as https://www.npmjs.com/package/safe-stable-stringify

This should also be checked in other parts of the code where we do similar things.

Please also add test cases for that.

Reply:
Good catch! Using safe-stable-stringify is probably overkill here. Since OpenTelemetry attributes only use string keys and primitive (or array of primitive) values, we don’t need to handle nested objects. I’ve pushed a commit with a simple stableStringify implementation.